File: core/src/main/java/com/illuzionzstudios/tab/controller/GroupController.java
package com.illuzionzstudios.tab.controller;

import com.illuzionzstudios.mist.Logger;
import com.illuzionzstudios.mist.config.ConfigSection;
import com.illuzionzstudios.mist.config.PluginSettings;
import com.illuzionzstudios.mist.config.YamlConfig;
import com.illuzionzstudios.mist.controller.PluginController;
import com.illuzionzstudios.mist.plugin.SpigotPlugin;
import com.illuzionzstudios.tab.CustomTab;
import com.illuzionzstudios.tab.components.loader.GroupLoader;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;

import java.util.HashMap;

/**
 * Control player tab groups
 */
public enum GroupController implements PluginController<SpigotPlugin> {
    INSTANCE;

    /**
     * All loaded groups
     */
    public HashMap<String, GroupLoader> groups = new HashMap<>();

    @Override
    public void initialize(SpigotPlugin plugin) {
        // Load groups
        YamlConfig config = PluginSettings.SETTINGS_FILE;

        // Loop through and create a loader for each section
        for (ConfigSection section : config.getSections("Tab.Groups")) {
            // Notify the column has been loaded
            Logger.info("Loaded tab group '" + section.getName() + "'");
            this.groups.put(section.getName().toLowerCase(), new GroupLoader(section));
        }
    }

    @Override
    public void stop(SpigotPlugin plugin) {
        this.groups.clear();
    }

    /**
     * Gets the highest group the player has
     *
     * @param player The player to check
     */
    public GroupLoader getGroup(Player player) {
        GroupLoader highest = null;

        // Weird permissions deop while doing check
        boolean wasOp = false;
        if (player.isOp()) {
            wasOp = true;
            player.setOp(false);
        }

        for (GroupLoader group : groups.values()) {
            // Has permission for group
            if (player.hasPermission(group.getPermission()) || group.getPermission().trim().equalsIgnoreCase("")) {
                int compare = Integer.compare(highest == null ? 0 : highest.getWeight(), group.getWeight());
                if (compare < 0) {
                    highest = group;
                }
            }
        }

        if (wasOp) {
            player.setOp(true);
        }

        return highest;
    }
}

File: core/src/main/kotlin/com/illuzionzstudios/tab/group/GroupController.kt
package com.illuzionzstudios.tab.group

import com.illuzionzstudios.mist.Logger
import com.illuzionzstudios.mist.config.YamlConfig
import com.illuzionzstudios.mist.controller.PluginController
import com.illuzionzstudios.mist.plugin.SpigotPlugin
import com.illuzionzstudios.mist.util.PlayerUtil
import org.bukkit.entity.Player
import java.util.*
import kotlin.collections.HashMap

object GroupController: PluginController {

    /**
     * All loaded group instances
     */
    val groups: MutableMap<String, Group> = HashMap()

    /**
     * A cache of assigned groups so we don't constantly have to do check for groups
     */
    val groupCache: MutableMap<UUID, Group> = HashMap()

    override fun initialize(plugin: SpigotPlugin) {
        YamlConfig.loadInternalYaml(plugin, "", "groups.yml")

        // Load groups
        GroupLoader().`object`.forEach {
            Logger.info("Loading group `" + it.id + "`")
            groups[it.id] = it
        }
    }

    override fun stop(plugin: SpigotPlugin) {
        groups.clear()
        groupCache.clear()
    }

    /**
     * Gets the highest group the player has
     *
     * @param player The player to check
     */
    fun getGroup(player: Player): Group? {
        // Find in cache first
//        if (groupCache.contains(player.uniqueId)) return groupCache[player.uniqueId]

        var highest: Group? = null
        for (group in groups.values) {
            // Has permission for group
            if (PlayerUtil.hasPerm(player, group.permission, true) || group.permission.trim().equals("", true)) {
                if ((highest?.weight ?: 0) < group.weight) {
                    highest = group
                }
            }
        }

//        groupCache[player.uniqueId] = highest ?: groups.values.first()
        return highest
    }
}